resource/kind.go

```diff
@@ -22,334 +18,162 @@ const (
-	if in == nil {
-		return fmt.Errorf("in io.Reader cannot be nil")
-	}
-	if into == nil {
-		return fmt.Errorf("into Object cannot be nil")
-	}
-	switch encoding {
-	case KindEncodingJSON:
-		return json.NewDecoder(in).Decode(&into)
-	case KindEncodingYAML:
-		return yaml.NewDecoder(in).Decode(&into)
+// Scope returns Schema.Scope() if Schema is non-nil, otherwise it returns an empty string
+func (k *Kind) Scope() SchemaScope {
+	if k.Schema == nil {
+		return ""
 	}
-	return fmt.Errorf("cannot unmarshal unknown content encoding '%s'", encoding)
+	return k.Schema.Scope()
 }
 
-// Write takes in any Object and outputs a kubernetes object JSON
-func (u *UntypedKind) Write(obj Object, out io.Writer, encoding KindEncoding) error {
-	m := make(map[string]any)
-	m["apiVersion"], m["kind"] = obj.GetObjectKind().GroupVersionKind().ToAPIVersionAndKind()
-	m["metadata"] = metav1.ObjectMeta{
-		Name:                       obj.GetName(),
-		GenerateName:               obj.GetGenerateName(),
-		Namespace:                  obj.GetNamespace(),
-		SelfLink:                   obj.GetSelfLink(),
-		UID:                        obj.GetUID(),
-		ResourceVersion:            obj.GetResourceVersion(),
-		Generation:                 obj.GetGeneration(),
-		CreationTimestamp:          obj.GetCreationTimestamp(),
-		DeletionTimestamp:          obj.GetDeletionTimestamp(),
-		DeletionGracePeriodSeconds: obj.GetDeletionGracePeriodSeconds(),
-		Labels:                     obj.GetLabels(),
-		Annotations:                obj.GetAnnotations(),
-		OwnerReferences:            obj.GetOwnerReferences(),
-		Finalizers:                 obj.GetFinalizers(),
-		ManagedFields:              obj.GetManagedFields(),
-	}
-	m["spec"] = obj.GetSpec()
-	for k, v := range obj.GetSubresources() {
-		m[k] = v
-	}
-	switch encoding {
-	case KindEncodingJSON:
-		return json.NewEncoder(out).Encode(m)
-	case KindEncodingYAML:
-		return yaml.NewEncoder(out).Encode(m)
-	default:
-		return fmt.Errorf("cannot marshal unknown content encoding '%s'", encoding)
+// ZeroValue returns the ZeroValue() function of the Schema, or an *UntypedObject if the Schema is nil.
+func (k *Kind) ZeroValue() Object {
+	if k.Schema == nil {
+		return &UntypedObject{}
 	}
+	return k.Schema.ZeroValue()
 }
 
-// NewTypedKind is a convenience function for creating a new instance of TypedKind.
-// It defaults all fields not required as arguments.
-func NewTypedKind[T Object](group, version, kind string, zeroObject T) *TypedKind[T] {
-	return &TypedKind[T]{
-		GVK: schema.GroupVersionKind{
-			Group:   group,
-			Version: version,
-			Kind:    kind,
-		},
-		PluralKind: fmt.Sprintf("%ss", strings.ToLower(kind)),
-		KindScope:  NamespacedScope,
-		ZeroObject: zeroObject,
+// Codec is a nil-safe way of accessing the Codecs map in the Kind.
+// It will return nil if the map key does not exist, or the key is explicitly set to nil.
+func (k *Kind) Codec(encoding KindEncoding) Codec {
```

Comment:
Question – do we want to redefine Read / Write methods on the Kind?

Something like:
```go
func (k Kind) Read(src io.Reader, enc KindEncoding) (Object, error) {
  c, ok := k.Codecs[enc]
  if !ok {
    return fmt.Errorf("no available codec for encoding: %s", enc)
  }

  res := k.Schema.ZeroValue()

  if err := c.Read(src, res); err != nil {
    return nil, err
  }

  return res, nil
}
```